quic/congestion_control/Copa2.cpp

```diff
@@ -0,0 +1,262 @@
+  }
+}
+
+// Switch to and from lossy mode
+void Copa2::manageLossyMode(folly::Optional<TimePoint> sentTime) {
+  if (!sentTime) {
+    // Loss happened and we don't know when. Be safe
+    lossyMode_ = true;
+    numAckedInLossCycle_ = 0;
+    numLostInLossCycle_ = 0;
+    lossCycleStartTime_ = Clock::now();
+    return;
+  }
+
+  auto numPktsInLossCycle = numAckedInLossCycle_ + numLostInLossCycle_;
+  if (*sentTime < lossCycleStartTime_) {
+    // Wait for at-least one RTT before declaring lossyMode_
+    return;
+  }
+  if (numPktsInLossCycle < 2 / lossToleranceParam_ && numLostInLossCycle_ < 2) {
+    // Second condition is needed in case there are losses, but not acks
+    return;
+  }
+  VLOG(5) << __func__ << " lossyMode=" << lossyMode_
+          << " num lost=" << numLostInLossCycle_
+          << " num acked=" << numAckedInLossCycle_ << " " << conn_;
+  // Cycle has ended. Take stock of the situation
+  DCHECK(numPktsInLossCycle > 0);
+  lossyMode_ = numLostInLossCycle_ >= numPktsInLossCycle * lossToleranceParam_;
+  numAckedInLossCycle_ = 0;
+  numLostInLossCycle_ = 0;
+  lossCycleStartTime_ = Clock::now();
+}
+
+void Copa2::onPacketLoss(const LossEvent& loss) {
+  VLOG(10) << __func__ << " lostBytes=" << loss.lostBytes
+           << " lostPackets=" << loss.lostPackets << " cwnd=" << cwndBytes_
+           << " inflight=" << conn_.lossState.inflightBytes << " " << conn_;
+  if (conn_.qLogger) {
+    conn_.qLogger->addCongestionMetricUpdate(
+        conn_.lossState.inflightBytes,
+        getCongestionWindow(),
+        kCongestionPacketLoss);
+  }
+  DCHECK(loss.largestLostPacketNum.has_value());
+  subtractAndCheckUnderflow(conn_.lossState.inflightBytes, loss.lostBytes);
+  if (loss.persistentCongestion) {
+    VLOG(10) << __func__ << " writable=" << getWritableBytes()
+             << " cwnd=" << cwndBytes_
+             << " inflight=" << conn_.lossState.inflightBytes << " " << conn_;
+    cwndBytes_ = conn_.transportSettings.minCwndInMss * conn_.udpSendPacketLen;
+    if (conn_.pacer) {
+      // TODO Which min RTT should we use?
+      conn_.pacer->refreshPacingRate(cwndBytes_, conn_.lossState.mrtt);
+    }
+    if (conn_.qLogger) {
+      conn_.qLogger->addCongestionMetricUpdate(
+          conn_.lossState.inflightBytes,
+          getCongestionWindow(),
+          kPersistentCongestion);
+    }
+  }
+
+  numLostInLossCycle_ += loss.lostPackets;
+  manageLossyMode(loss.largestLostSentTime);
+}
+
+void Copa2::onPacketAcked(const AckEvent& ack) {
+  DCHECK(ack.largestAckedPacket.has_value());
+  subtractAndCheckUnderflow(conn_.lossState.inflightBytes, ack.ackedBytes);
+  minRTTFilter_.Update(
+      conn_.lossState.lrtt,
+      std::chrono::duration_cast<microseconds>(ack.ackTime.time_since_epoch())
+          .count());
+  bytesAckedInCycle_ += ack.ackedBytes;
+  for (const auto& ackPkt : ack.ackedPackets) {
+    appLimitedInCycle_ = appLimitedInCycle_ || ackPkt.isAppLimited;
```

Comment:
This will always be false, right? ackPkt.isAppLimit needs the CC to be able to answer isAppLimit()
I guess you decided to add this code in case in the future app limited tracking will be supported in copa2?

Reply:
I thought this worked :smile: 

Ok, I want to implement this right. The semantic I want is that if the sender was app limited (did not send packets even though cwnd and pacer allowed it) at any point when the packets in our cycle were sent, we should not decrease cwnd.

What is the semantic for isAppLimited being called? The comments say that it is called when it starts and stops being app limited, but there is no parameter. Is it like a switch that turns on and off?

I assume isAppIdle means the app is not sending bytes, but the socket buffer may be non-empty?

Reply:
It is set when upon ack, by whether the packet was set to be app limited during sent time: https://github.com/facebookincubator/mvfst/blob/a3c78f784812293a20c790eb8112c1a153e8ea5b/quic/state/AckHandlers.cpp#L206

And the sent packet's isAppLimited is set after we write to socket, we query the congestion controller if it thinks the connection state is app limited: https://github.com/facebookincubator/mvfst/blob/96bf643f4f48655953c2972b57605370b3607e64/quic/api/QuicTransportFunctions.cpp#L692

And this is where we set the app limited to the congestion controller:
https://github.com/facebookincubator/mvfst/blob/c1223a2f78ba24505b34510322d7dbc7e9838845/quic/api/QuicTransportBase.cpp#L2773

The condition is based on bbr v1 specs. We may need to refactor this code a little bit if you want Copa2 to have a different condition to set it.

Reply:
AppIdle is when we do not have request streams in the connection: https://github.com/facebookincubator/mvfst/blob/ac3eb576cbf8d277bc49a79eb02a27c39d94db58/quic/state/QuicStreamManager.cpp#L536

Reply:
Oh, forgot to mention, there is no API to unset app-limited, only an API to set it. That's because we only use it in BBR now and BBR's bandwidth sampler will unset app-limited itself.